# Pandas kutubxonasiga doir bilimlaringizni mustaxkamlash uchun mashqlar!

Hurmatli o'quvchi, bilimingizni yanada mustaxkamlashingiz va chuqurroq ko'nikmalarga ega bo'lishingiz uchun sizga topshiriqlar tayyorladik. Sizga yengillik yaratish maqsadida topshiriqlarni `📌Mavzulashtirdik`

## `📌Pandas kutubxonasini yuklash`

**1.** Pandas kutubxonasini `pd` nomi ostida yuklang va pandas kutubxonasini chop eting.

In [1]:
import pandas as pd

**2.** Versiya va konfiguratsiyani tekshiring.

In [2]:
pd.__version__

'2.0.3'

**3.** Pandas kutubxonasida mavjud bo'lgan barcha versiya ma'lumotlarini chop eting.

In [3]:
pd.show_versions

<function pandas.util._print_versions.show_versions(as_json: 'str | bool' = False) -> 'None'>

## `📌DataFrame: Asoslar`

DataFrames-da ma'lumotlarni tanlash(selecting), saralash(sorting), qo'shish(adding) va yig'ish(aggregating) uchun bir nechta asosiy mashqlar.

💡Eslatma: NumPy kutubxonasini import qilishni unutmang:
``` python
import numpy as np
```

Quyidagi Python lug'ati(dictionary) `data` va Python ro'yxati(list) `labels` ni ko'rib chiqing:


``` python
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
```



In [4]:
import numpy as np

In [5]:
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)

**4.** Ushbu `lug'at`(dictionary) yordamida `data` dan `labels` indeksiga ega DataFrame yarating.

In [6]:
import numpy as np

data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)

**5.** Ushbu DataFrame va uning ma'lumotlari haqidagi asosiy ma'lumotlarning qisqacha mazmunini ko'rsating.

`💡Ishora:` DataFrame-da chaqirilishi mumkin bo'lgan yagona usul mavjud.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   animal    10 non-null     object 
 1   age       8 non-null      float64
 2   visits    10 non-null     int64  
 3   priority  10 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


In [8]:
df.describe()

age     visits
count  8.000000  10.000000
mean   3.437500   1.900000
std    2.007797   0.875595
min    0.500000   1.000000
25%    2.375000   1.000000
50%    3.000000   2.000000
75%    4.625000   2.750000
max    7.000000   3.000000

**6.** DataFrame `df` dastlabki 3 qatorini qaytaring.

In [9]:
df.head(3)

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

In [10]:
df.iloc[:3]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

**7.** DataFrame `df`-dan faqat `animal` va `age` ustunlarini tanlang.

In [11]:
df.loc[:, ['animal', 'age']]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [12]:
df[['animal', 'age']]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

**8.** `[3, 4, 8]` va `['animal', 'age']` qatorlaridagi ma'lumotlarni tanlang.

In [13]:
df.loc[df.index[[3, 4, 8]], ['animal', 'age']]

animal  age
d    dog  NaN
e    dog  5.0
i    dog  7.0

**9.** Faqat `visits` soni 3 tadan ko'p bo'lgan qatorlarni tanlang.

In [14]:
df[df['visits']>2]

animal  age  visits priority
b    cat  3.0       3      yes
d    dog  NaN       3      yes
f    cat  2.0       3       no

10. Yosh yo'qolgan qatorlarni tanlang, ya'ni `NaN`

In [15]:
df[df['age'].isnull()]

animal  age  visits priority
d    dog  NaN       3      yes
h    cat  NaN       1      yes

**11.** `animal` qatoridagi `cat` va `age` qatoridagi `3` dan kichik bo'lgan umumiy qatorlarni tanlang.

In [16]:
df[(df['animal']=='cat') & (df['age']<3)]

animal  age  visits priority
a    cat  2.5       1      yes
f    cat  2.0       3       no

**12.** `age` 2 va 4 (shu jumladan) orasidagi qatorlarni tanlang.

In [17]:
df[df['age'].between(2, 4)]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
f    cat  2.0       3       no
j    dog  3.0       1       no

**13.** `f` qatoridagi yoshni `1,5` ga o'zgartiring.

In [18]:
df.loc['f', 'age'] = 1.5

**14.** `df` yordamida `visits` qatorining yig'indisini hisoblang.

In [19]:
df['visits'].sum()

19

**15.** Har bir `animal` uchun o'rtacha yoshni `df` da hisoblang.

In [21]:
df.groupby('animal')['age'].mean()

animal
cat      2.333333
dog      5.000000
snake    2.500000
Name: age, dtype: float64

**16.** `df` yordamida har bir turdagi `animal` sonini hisoblang.

In [22]:
df['animal'].value_counts()

animal
cat      4
dog      4
snake    2
Name: count, dtype: int64

**17.** `animal` ustunida, `snake` so'zini `python` so'ziga o'zgartiring.

In [23]:
df['animal'] = df['animal'].replace('snake', 'python')

## `📌DataFrame: Murakkab`

### Bir oz qiyinroq: to'g'ri javob olish uchun ikki yoki undan ortiq usullarni birlashtirish kerak bo'lishi mumkin

Oldingi bo'lim DataFrame-ning asosiy, ammo muhim operatsiyalari bo'yicha vazifalarni jamlagan edi. Quyida ma'lumotlarni kesishingiz kerak bo'lgan mashqlar mavjud.

**18.** Sizda butun sonlardan iborat `A` ustunli DataFrame `df` mavjud. Misol uchun:
```python
df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
```

Yuqoridagi qator bilan bir xil butun sonni o'z ichiga olgan qator yarating.


💡Sizda quyidagi qiymatlarni o'z ichiga olgan ustun paydo bo'lishi kerak:

```python
1, 2, 3, 4, 5, 6, 7
```

In [24]:
df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})

df.loc[df['A'].shift() != df['A']]
df.drop_duplicates(subset='A')

A
0  1
1  2
3  3
4  4
5  5
8  6
9  7

**19.** Raqamli qiymatlardan iborat DataFrame berildi

```python
df = pd.DataFrame(np.random.random(size=(5, 3))) # float qiymatlarining 5x3 o'lchovi
```

Qatordagi har bir elementdan satrning o'rtacha qiymatini qanday ayirish mumkin?

In [25]:
df = pd.DataFrame(np.random.random(size=(5, 3)))

In [26]:
df

0         1         2
0  0.574096  0.347228  0.800196
1  0.406086  0.222576  0.321261
2  0.930577  0.692897  0.502329
3  0.154922  0.115439  0.704236
4  0.212337  0.056450  0.477688

In [27]:
df.sub(df.mean(axis=1), axis=0)

0         1         2
0  0.000256 -0.226612  0.226356
1  0.089445 -0.094065  0.004620
2  0.221976 -0.015704 -0.206273
3 -0.169944 -0.209427  0.379370
4 -0.036488 -0.192375  0.228863

**20.** Sizda 10 ta haqiqiy raqamlar ustunidan iborat DataFrame mavjud:

``` python
df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))
```
Qaysi sonlar ustunining yig'indisi eng kichik ekanini hisoblang. U ustunning yorlig'ini qaytaring.

In [28]:
df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))

df.sum().idxmin()

'i'

**21.** Quyida butun sonlardan iborat DataFrame `df` mavjud. Har bir satrga  5 ta yozuv `NaN` qiymatlari kiritilgan.

DataFrame ning har bir qatori uchun uchinchi `NaN` qiymatini o'z ichiga olgan ustunni toping.

Ustun yorliqlarini qaytarishingiz kerak: `e, c, d, h, d`

In [29]:
nan = np.nan
data = [[0.04,  nan,  nan, 0.25,  nan, 0.43, 0.71, 0.51,  nan,  nan],
        [ nan,  nan,  nan, 0.04, 0.76,  nan,  nan, 0.67, 0.76, 0.16],
        [ nan,  nan, 0.5 ,  nan, 0.31, 0.4 ,  nan,  nan, 0.24, 0.01],
        [0.49,  nan,  nan, 0.62, 0.73, 0.26, 0.85,  nan,  nan,  nan],
        [ nan,  nan, 0.41,  nan, 0.05,  nan, 0.61,  nan, 0.48, 0.68]]

columns = list('abcdefghij')

df = pd.DataFrame(data, columns=columns)

(df.isnull().cumsum(axis=1) == 3).idxmax(axis=1)

0    e
1    c
2    d
3    h
4    d
dtype: object

**22.** DataFrame `grps` guruhlari ustuniga va `vals` butun qiymatlari ustuniga ega:

```python
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})
```
Har bir `grps` uchun uchta eng katta qiymatning yig‘indisini toping.

💡Javobni quyidagicha yakunlashingiz kerak:
```
grps
a    409
b    156
c    345
```

     'a   a  a b  b  c a a  b  c  c   c  b  b  c'
      12,345,3,1,45,14,4,52,54,23,235,21,57,3,87

In [35]:
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})

df.groupby('grps')['vals'].nlargest(3).groupby('grps').sum()

grps
a    409
b    156
c    345
Name: vals, dtype: int64

In [36]:
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})

# Javobni bu yerga yozing!
df.groupby('grps')['vals'].nlargest(3).groupby('grps').sum()

grps
a    409
b    156
c    345
Name: vals, dtype: int64

**28.** Quyida tuzilgan DataFrame `df` ikkita butun son ustuniga ega 'A' va 'B'. "A" dagi qiymatlar 1 dan 100 gacha (shu jumladan).

“A” dagi har bir 10 ta ketma-ket butun sonlar guruhi uchun (ya’ni `(0, 10]`, `(10, 20]`, ...) 'B' ustunidagi mos qiymatlar yig'indisini hisoblang.

Javob quyidagicha bo'lishi kerak:

```
A
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
```

In [37]:
df = pd.DataFrame(np.random.RandomState(8765).randint(1, 101, size=(100, 2)), columns = ["A", "B"])

# Javobni bu yerga yozing!
df.groupby(pd.cut(df['A'], np.arange(0, 101, 10)))['B'].sum()

A
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
Name: B, dtype: int64

**29.** 'X' butun son ustuni bo'lgan DataFrame:
```python
df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})
```

Har bir qiymat uchun farqni oldingi nolga qaytaring.

Sizda natija quyidagicha bo'lishi kerak:

```
[1, 2, 0, 1, 2, 3, 4, 0, 1, 2]
```

Buni yangi `Y` ustuni deb nomlang.

In [39]:
# Asl DataFrame
df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})

# Yangi ustun uchun bo'sh ro'yxat
distance_from_last_zero = []
last_zero_index = -1

# X ustunidagi qiymatlar bo'yicha aylanish
for i, value in enumerate(df['X']):
    if value == 0:
        last_zero_index = i
    distance_from_last_zero.append(i - last_zero_index)

# Y ustunini qo'shish
df['Y'] = distance_from_last_zero

print(df)

   X  Y
0  7  1
1  2  2
2  0  0
3  3  1
4  4  2
5  2  3
6  5  4
7  0  0
8  3  1
9  4  2


##`📌Ma'lumotlarni tozalash`

Bu har doim sodir bo'ladi: kimdir sizga noto'g'ri tuzilgan satrlar, ro'yxatlar va yetishmayotgan ma'lumotlarni o'z ichiga olgan ma'lumotlarni beradi. Tahlilni davom ettirish uchun uni qanday tartibga keltirasiz?

**30.** Quyidagi jumboqni tasavvur qilishi uchun ushbu misolni DataFrame sifatida oling:

```python
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm',
                               'Budapest_PaRis', 'Brussels_londOn'],
              'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
              'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
                   'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )',
                               '12. Air France', '"Swiss Air"']})
```
Tartiblaganingizdan so'ng u quyidagi ko'rinishga ega bo'lishi kerak:

```
            From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"
```


In [61]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm',
                               'Budapest_PaRis', 'Brussels_londOn'],
                   'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
                   'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
                   'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )',
                               '12. Air France', '"Swiss Air"']})

# FlightNumber ustunidagi etishmayotgan qiymatlarni to'ldirish
df['FlightNumber'] = df['FlightNumber'].interpolate().astype(int)

# From_To ustunini ikkita alohida ustunga bo'lish
df[['From', 'To']] = df['From_To'].str.split('_', expand=True)

# Shahar nomlarini to'g'ri formatda yozish
df['From'] = df['From'].str.capitalize()
df['To'] = df['To'].str.capitalize()

# RecentDelays ustunidagi qiymatlarni alohida ustunlarga ajratish
delays = df['RecentDelays'].apply(pd.Series)
delays.columns = [f'Delay_{i+1}' for i in range(delays.shape[1])]

# Yakuniy DataFrame'ni to'g'ri formatda tuzish
df = df.drop(['From','To'], axis=1)
df = pd.concat([df, delays], axis=1)

print(df)

            From_To  FlightNumber  RecentDelays              Airline  Delay_1  \
0      LoNDon_paris         10045      [23, 47]               KLM(!)     23.0   
1      MAdrid_miLAN         10055            []    <Air France> (12)      NaN   
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )     24.0   
3    Budapest_PaRis         10075          [13]       12. Air France     13.0   
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"     67.0   

   Delay_2  Delay_3  
0     47.0      NaN  
1      NaN      NaN  
2     43.0     87.0  
3      NaN      NaN  
4     32.0      NaN  


**31.** **FlightNumber** ustunidagi ba'zi qiymatlar yo'q. Ular `NaN` ko'rinishida yozilgan. Bu raqamlar har bir qatorda 10 ga ko'payishi kerak, shuning uchun 10055 va 10075 ni qo'yish kerak. Ushbu yetishmayotgan raqamlarni to'ldirish uchun `df` ni o'zgartiring va ustunni butun son ustuniga aylantiring.

In [62]:
df['FlightNumber'] = df['FlightNumber'].interpolate().astype(int)
df['FlightNumber'] = df['FlightNumber'] * 10

**32.** **From\_To** ustunini ikkita alohida ustun ko'rinishida yozing.                                          

To'g'ri qiymatlarga ega `temp` deb nomlangan yangi Temporary DataFrame berish uchun har bir satrni pastki chiziq chegarasiga ajrating.                                                                     Ushbu Temporary DataFrame ustunlarini `From` va `To` deb nomlang.

In [63]:
temp = df['From_To'].str.split('_', expand=True)
temp.columns = ['From', 'To']

**34.** Temporary DataFrame `temp` da shahar nomlari betartib yozilgan.Satrlarning faqat birinchi harfi katta yozilishi kerak.
                                                                         (Masalan, "londON" "London" bo'lishi kerak.)

In [64]:
temp['From'] = temp['From'].str.capitalize()
temp['To'] = temp['To'].str.capitalize()


**33**. `Airline` ustunida aviakompaniya nomlari atrofida qoʻshimcha tinish belgilari va belgilar paydo boʻlganini koʻrishingiz mumkin. Faqat aviakompaniya nomini qoldiring. Masalan, ʻ'(British Airways. )'' `British Airways` ko'rinishida yozilishi kerak.

In [66]:
# Airline ustunidagi ortiqcha belgilarni olib tashlash
df['Airline'] = df['Airline'].str.replace(r'[^\w\s]', '', regex=True).str.strip()

df['Airline'].head()

0                KLM
1      Air France 12
2    British Airways
3      12 Air France
4          Swiss Air
Name: Airline, dtype: object

**34.** `df` dan **From_To** ustunini o'chiring va Temporary DataFrame `temp` ustunini qo'shing.

In [67]:
df.drop('From_To', axis=1, inplace=True)
df = pd.concat([df, temp], axis=1)
print(df)

   FlightNumber  RecentDelays          Airline  Delay_1  Delay_2  Delay_3  \
0        100450      [23, 47]              KLM     23.0     47.0      NaN   
1        100550            []    Air France 12      NaN      NaN      NaN   
2        100650  [24, 43, 87]  British Airways     24.0     43.0     87.0   
3        100750          [13]    12 Air France     13.0      NaN      NaN   
4        100850      [67, 32]        Swiss Air     67.0     32.0      NaN   

       From         To  
0    London      Paris  
1    Madrid      Milan  
2    London  Stockholm  
3  Budapest      Paris  
4  Brussels     London  


**34**. `RecentDelays` ustunida qiymatlar DataFrame-ga ro'yxat sifatida kiritilgan. Sizning vazifangiz birinchi qiymatni o'z ustunida, har bir ikkinchi qiymatni o'z ustunida va qolgan qiymatlar ham ustunga mos holda chop etish. Agar N-chi qiymat bo'lmasa, qiymat `NaN` deb nomlanishi kerak.

Roʻyxatlar qatorini `delays` nomli DataFrame ichiga kengaytiring, `delay_1`, `delay_2` va qolgan ustunlarni ham nomlarini oʻzgartiring va `df` dagi keraksiz `RecentDelays` ustunini `delays` bilan almashtiring.

In [68]:
delays = df['RecentDelays'].apply(pd.Series)
delays.columns = [f'Delay_{i+1}' for i in range(delays.shape[1])]

# Yakuniy DataFrame'ni to'g'ri formatda tuzish
df = df.drop(['RecentDelays'], axis=1)
df = pd.concat([df, delays], axis=1)

print(df)


   FlightNumber          Airline  Delay_1  Delay_2  Delay_3      From  \
0        100450              KLM     23.0     47.0      NaN    London   
1        100550    Air France 12      NaN      NaN      NaN    Madrid   
2        100650  British Airways     24.0     43.0     87.0    London   
3        100750    12 Air France     13.0      NaN      NaN  Budapest   
4        100850        Swiss Air     67.0     32.0      NaN  Brussels   

          To  Delay_1  Delay_2  Delay_3  
0      Paris     23.0     47.0      NaN  
1      Milan      NaN      NaN      NaN  
2  Stockholm     24.0     43.0     87.0  
3      Paris     13.0      NaN      NaN  
4     London     67.0     32.0      NaN  


Natija quyidagi ko'rinishda bo'lishi kerak:
```
   FlightNumber          Airline      From         To  delay_1  delay_2  delay_3
0         10045              KLM    London      Paris     23.0     47.0      NaN
1         10055       Air France    Madrid      Milan      NaN      NaN      NaN
2         10065  British Airways    London  Stockholm     24.0     43.0     87.0
3         10075       Air France  Budapest      Paris     13.0      NaN      NaN
4         10085        Swiss Air  Brussels     London     67.0     32.0      NaN
```

##`📌Minesweeper`


### Minesweeper tarmog'ida xavfsiz kvadratlar uchun raqamlar yaratamiz!

💡 Agar siz Windowsning avvalgi versiyalaridan foydalangan bo'lsangiz, Minesweeper bilan tanishsiz:
- https://en.wikipedia.org/wiki/Minesweeper_(video_game)

Agar siz o'yin bilan tanish bo'lmasangiz, kvadratchalar to'plamini tasavvur qiling: bu kvadratlarning ba'zilari mineni yashiradi. Agar siz mineni bossangiz, darhol yo'qotasiz. Agar siz xavfsiz kvadratni bossangiz, darhol qo'shni bo'lgan kvadratchalarda qancha mine borligini ko'rsatadigan raqamni ko'rasiz. O'yinning maqsadi to'rdagi mine bo'lmagan barcha kvadratlarni ochishdir.

Ushbu bo'limda biz Minesweeper o'yini uchun kerakli ma'lumotlarni o'z ichiga olgan DataFrame yaratamiz: kvadratlarning koordinatalari, kvadratda mine bormi va qo'shni kvadratlarda topilgan minelar sonini aniqlaymiz!

**35**. Tasavvur qiling, biz Minesweeper o'yinini 5 dan 4 gacha bo'lgan panjarada o'ynayapmiz, ya'ni.
```
X = 5
Y = 4
```
Boshlash uchun ikkita ustunli DataFrame `df` yarating, bu tarmoq uchun har bir koordinatani o'z ichiga olgan `'x'` va `'y'`. Ya'ni DataFrame ishga tushishi kerak:
```
    x y
0 0 0
1 0 1
2 0 2
```

In [70]:

# X va Y o'lchamlarini aniqlash
X = 5
Y = 4

# Bo'sh list yaratish
data = []

# Koordinatalarni yaratish va listga qo'shish
for x in range(X):
    for y in range(Y):
        data.append({'x': x, 'y': y})

# DataFrame yaratish
df = pd.DataFrame(data)

# Natijani chiqarish
print(df)

    x  y
0   0  0
1   0  1
2   0  2
3   0  3
4   1  0
5   1  1
6   1  2
7   1  3
8   2  0
9   2  1
10  2  2
11  2  3
12  3  0
13  3  1
14  3  2
15  3  3
16  4  0
17  4  1
18  4  2
19  4  3


**36**. Ushbu DataFrame `df` uchun yangi nol (safe) va birliklar (mine) ustunini yarating. Har bir joyda mine sodir bo'lish ehtimoli 0,4 bo'lishi kerak.

In [71]:
df['safe'] = (np.random.random(size=len(df)) < 0.6).astype(int)
df['mine'] = (np.random.random(size=len(df)) < 0.4).astype(int)

**37**. Endi ushbu DataFrame uchun `adjacent` nomli yangi ustun yarating. Ushbu ustunda tarmoqdagi qo'shni kvadratlarda topilgan minelar soni bo'lishi kerak.

(Masalan, `(0, 0)` koordinatasi uchun yozuv bo'lgan birinchi qator uchun `(0, 1)`, `(1, 0)` va `(1) koordinatalarida qancha mine borligini hisoblang. , 1)`.)

In [80]:
adjacent = []

# Har bir kvadrat uchun qo'shni koordinatalarda minelar sonini hisoblash
for i, row in df.iterrows():
    # Har bir koordinata uchun qo'shni koordinatalarni aniqlash
    neighbors = [(row['x'] + dx, row['y'] + dy) for dx in [-1, 0, 1] for dy in [-1, 0, 1]
                 if (0 <= row['x'] + dx < X) and (0 <= row['y'] + dy < Y) and not (dx == 0 and dy == 0)]

    # Qo'shni koordinatalarda topilgan minelar sonini hisoblash
    mine_count = sum(df[(df['x'] == x) & (df['y'] == y)]['mine'].values[0] for x, y in neighbors)

    adjacent.append(mine_count)

# adjacent ustunini DataFramega qo'shish
df['adjacent'] = adjacent

df

x  y  safe  mine  adjacent
0   0  0     1     1         1
1   0  1     1     0         3
2   0  2     0     0         3
3   0  3     0     1         2
4   1  0     1     1         2
5   1  1     1     0         5
6   1  2     1     1         4
7   1  3     1     1         3
8   2  0     0     0         2
9   2  1     0     1         3
10  2  2     1     1         4
11  2  3     1     0         4
12  3  0     0     0         2
13  3  1     0     0         3
14  3  2     1     0         3
15  3  3     1     1         1
16  4  0     0     1         0
17  4  1     1     0         1
18  4  2     1     0         1
19  4  3     1     0         1

**38**. Mineni o'z ichiga olgan DataFrame qatorlari uchun `adjacent` ustunidagi qiymatni `NaN` ga o'zgartiring.

In [81]:
df.loc[df['mine'] == 1, 'adjacent'] = np.nan

**39**. DataFrame-ni `adjacent` minelar sonining panjarasiga aylantiring: ustunlar `x` koordinatasi, qatorlar `y` koordinatasi.

In [82]:
df.pivot(index='y', columns='x', values='adjacent')


x    0    1    2    3    4
y                         
0  NaN  NaN  2.0  2.0  NaN
1  3.0  5.0  NaN  3.0  1.0
2  3.0  NaN  NaN  3.0  1.0
3  NaN  NaN  4.0  NaN  1.0

##📌BONUS!

O'zingiz javob berishni istagan savolni kiriting va unga javob bering!

In [83]:
my_dict = {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}

# 'a', 'c' va 'e' kalitlarini o'chirish
deleted_keys = my_dict.pop('a'), my_dict.pop('c'), my_dict.pop('e')

# O'chirilgan kalitlarni print qilish
print(deleted_keys)

(1, 3, 5)
